code/game/objects/structures/displaycase.dm

```diff
@@ -305,92 +311,101 @@
-		if(added_roundstart)
-			is_locked = !is_locked
-			to_chat(user, span_notice("You [!is_locked ? "un" : ""]lock the case."))
+/obj/structure/displaycase/trophy/dump()
+	if (showpiece)
+		if(holographic_showpiece)
+			visible_message(span_danger("The [showpiece] fizzles and vanishes!"))
+			do_sparks(1, FALSE, src)
+			QDEL_NULL(showpiece)
+			holographic_showpiece = FALSE
 		else
-			to_chat(user, span_warning("The lock is stuck shut!"))
-		return
+			. = ..()
+		placer_key = ""
+		trophy_message = null
 
-	if(is_locked)
-		to_chat(user, span_warning("The case is shut tight with an old-fashioned physical lock. Maybe you should ask the curator for the key?"))
+/obj/structure/displaycase/trophy/insert_showpiece(obj/item/new_showpiece, mob/user)
+	if(..())
 		return
-
-	if(!added_roundstart)
-		to_chat(user, span_warning("You've already put something new in this case!"))
-		return
-
-	if(is_type_in_typecache(W, GLOB.blacklisted_cargo_types))
-		to_chat(user, span_warning("The case rejects the [W]!"))
-		return
-
-	for(var/a in W.get_all_contents())
-		if(is_type_in_typecache(a, GLOB.blacklisted_cargo_types))
-			to_chat(user, span_warning("The case rejects the [W]!"))
-			return
-
-	if(user.transferItemToLoc(W, src))
-
-		if(showpiece)
-			to_chat(user, span_notice("You press a button, and [showpiece] descends into the floor of the case."))
-			QDEL_NULL(showpiece)
-
-		to_chat(user, span_notice("You insert [W] into the case."))
-		showpiece = W
-		added_roundstart = FALSE
-		update_appearance()
-
+	if(showpiece == new_showpiece)
 		placer_key = user.ckey
 
-		trophy_message = W.desc //default value
-
-		var/chosen_plaque = tgui_input_text(user, "What would you like the plaque to say? Default value is item's description.", "Trophy Plaque", trophy_message)
-		if(chosen_plaque)
-			if(user.Adjacent(src))
-				trophy_message = chosen_plaque
-				to_chat(user, span_notice("You set the plaque's text."))
-			else
-				to_chat(user, span_warning("You are too far to set the plaque's text!"))
+/obj/structure/displaycase/trophy/proc/toggle_historian_mode(mob/user)
+	historian_mode = !historian_mode
+	to_chat(user, span_notice("You [!historian_mode ? "disable" : "enable"] the case's historian mode."))
+	playsound(src, 'sound/machines/twobeep.ogg', 50)
+	SStgui.update_uis(src)
 
-		SSpersistence.SaveTrophy(src)
-		return TRUE
+/obj/structure/displaycase/trophy/toggle_lock(mob/user)
+	..()
+	SStgui.close_uis(src)
 
-	else
-		to_chat(user, span_warning("\The [W] is stuck to your hand, you can't put it in the [src.name]!"))
+/obj/structure/displaycase/trophy/ui_data(mob/user)
+	var/list/data = list()
+	data["historian_mode"] = historian_mode
+	data["holographic_showpiece"] = holographic_showpiece
+	data["max_length"] = MAX_PLAQUE_LEN
+	data["has_showpiece"] = showpiece ? TRUE : FALSE
+	if(showpiece)
+		data["showpiece_name"] = capitalize(format_text(showpiece.name))
+		var/base64 = icon2base64(getFlatIcon(showpiece,  no_anim=TRUE))
```

Comment:
Doing this every `ui_data` is extremely expensive. Cache the result of this somewhere and update it every time showpiece changes.

Reply:
What creating a new common parent type for trophy and forsale (`/obj/structure/displaycase/interactible` ?) and storing it on a var in there be sufficient enough, or should i use a spritesheet?

Reply:
Oh wait. I could just make the base displaycase also have a tgUI of their own, for consistency. Though if I did that, I would move trophy_message back to the outermost layer, and create plaques for the roundstart display cases.